### PR TITLE
Adding validation of studies removing custom events currently in their schedule

### PR DESCRIPTION
**Pull request by jnelsonjava**
For Bridge-3135.

- Preventing removal of a custom event from a study if that event is being used by a schedule.

### CODE REVIEW

File: src/main/java/org/sagebionetworks/bridge/services/StudyService.java [hunk 1]
```diff
@@ -350,4 +355,15 @@ private Study phaseTransition(String appId, String studyId, StudyPhase targetPha
         
         return study;
     }
+    
+    private Set<String> getCustomEventIdsFromSchedule(String appId, String scheduleGuid) {
+        Schedule2 schedule = scheduleService.getSchedule(appId, scheduleGuid);
```
Comment (alxdarksage) [inline src/main/java/org/sagebionetworks/bridge/services/StudyService.java]: A study can not have a scheduleGuid, so you can't guarantee it won't be null here.

File: src/main/java/org/sagebionetworks/bridge/services/StudyService.java [hunk 2]
```diff
@@ -350,4 +355,15 @@ private Study phaseTransition(String appId, String studyId, StudyPhase targetPha
...
+        Set<String> existingEventIds = new HashSet<>();
+        if (schedule != null) {
+            existingEventIds = schedule.getSessions().stream()
+                    .flatMap(session -> session.getStartEventIds().stream())
+                    .filter(s -> s.startsWith("custom")).collect(Collectors.toSet());
```
Comment (alxdarksage) [inline src/main/java/org/sagebionetworks/bridge/services/StudyService.java]: I would add the colon: s.startsWith("colon:")

File: src/main/java/org/sagebionetworks/bridge/validators/StudyValidator.java [hunk 3]
```diff
@@ -141,10 +146,15 @@ public void validate(Object object, Errors errors) {
...
+            scheduledCustomEventIds.remove(customEvent.getEventId());
             errors.popNestedPath();
         }
         if (uniqueIds.size() > 0 && (uniqueIds.size() != study.getCustomEvents().size())) {
-            errors.rejectValue(CUSTOM_EVENTS_FIELD, "cannot contain duplidate event IDs");
+            errors.rejectValue(CUSTOM_EVENTS_FIELD, "cannot contain duplicate event IDs");
+        }
+        if (!scheduledCustomEventIds.isEmpty()) {
+            errors.rejectValue(CUSTOM_EVENTS_FIELD, 
+                    String.format("cannot remove custom events currently used in a schedule: %s", scheduledCustomEventIds.toString()));
```
Comment (alxdarksage) [inline src/main/java/org/sagebionetworks/bridge/validators/StudyValidator.java]: You need to verify that the existing scheduleCustomEventIds are all in the study...but there could be none and that's okay (schedules can be saved with no sessions). Adding is okay here. Just not removing.

**Review by alxdarksage — CHANGES_REQUESTED** [review]

File: src/main/java/org/sagebionetworks/bridge/services/StudyService.java [hunk 4]
```diff
@@ -350,4 +355,17 @@ private Study phaseTransition(String appId, String studyId, StudyPhase targetPha
         
         return study;
     }
+    
+    private Set<String> getCustomEventIdsFromSchedule(String appId, String scheduleGuid) {
+        Set<String> existingEventIds = new HashSet<>();
+        if (scheduleGuid != null) {
+            Schedule2 schedule = scheduleService.getSchedule(appId, scheduleGuid);
+            if (schedule != null) {
```
Comment (jnelsonjava) [inline src/main/java/org/sagebionetworks/bridge/services/StudyService.java]: This null check is unnecessary.

File: src/main/java/org/sagebionetworks/bridge/validators/StudyValidator.java [hunk 5]
```diff
@@ -141,10 +146,16 @@ public void validate(Object object, Errors errors) {
...
             errors.popNestedPath();
         }
         if (uniqueIds.size() > 0 && (uniqueIds.size() != study.getCustomEvents().size())) {
-            errors.rejectValue(CUSTOM_EVENTS_FIELD, "cannot contain duplidate event IDs");
+            errors.rejectValue(CUSTOM_EVENTS_FIELD, "cannot contain duplicate event IDs");
+        }
+        // Any EventIds not removed from the set are on a schedule but missing from the validated study
+        if (!protectedCustomEventIds.isEmpty()) {
+            errors.rejectValue(CUSTOM_EVENTS_FIELD, 
+                    String.format("cannot remove custom events currently used in a schedule: %s", protectedCustomEventIds.toString()));
```
Comment (jnelsonjava) [inline src/main/java/org/sagebionetworks/bridge/validators/StudyValidator.java]: Look up comma space joiner

File: src/main/java/org/sagebionetworks/bridge/services/StudyService.java [hunk 6]
```diff
@@ -350,4 +355,17 @@ private Study phaseTransition(String appId, String studyId, StudyPhase targetPha
         
         return study;
     }
+    
+    private Set<String> getCustomEventIdsFromSchedule(String appId, String scheduleGuid) {
```
Comment (jnelsonjava) [inline src/main/java/org/sagebionetworks/bridge/services/StudyService.java]: Make it protected to test in service tests.

File: src/test/java/org/sagebionetworks/bridge/validators/StudyValidatorTest.java [hunk 7]
```diff
@@ -285,7 +295,22 @@ public void customEvents_duplicated() {
         study.getCustomEvents().add(event);
         study.getCustomEvents().add(event);
         
-        assertValidatorMessage(INSTANCE, study, CUSTOM_EVENTS_FIELD, "cannot contain duplidate event IDs");
+        assertValidatorMessage(validator, study, CUSTOM_EVENTS_FIELD, "cannot contain duplicate event IDs");
+    }
+
```
Comment (jnelsonjava) [inline src/test/java/org/sagebionetworks/bridge/validators/StudyValidatorTest.java:300]: Add test for empty cases just to make sure it's not going to break anything.

File: src/main/java/org/sagebionetworks/bridge/services/StudyService.java (hunk 6 above)
Comment (jnelsonjava) [reply src/main/java/org/sagebionetworks/bridge/services/StudyService.java]: Done

File: src/main/java/org/sagebionetworks/bridge/services/StudyService.java (hunk 4 above)
Comment (jnelsonjava) [reply src/main/java/org/sagebionetworks/bridge/services/StudyService.java]: Removed

File: src/main/java/org/sagebionetworks/bridge/validators/StudyValidator.java (hunk 5 above)
Comment (jnelsonjava) [reply src/main/java/org/sagebionetworks/bridge/validators/StudyValidator.java]: Swapped to joiner instead of toString

File: src/test/java/org/sagebionetworks/bridge/validators/StudyValidatorTest.java (hunk 7 above)
Comment (jnelsonjava) [reply src/test/java/org/sagebionetworks/bridge/validators/StudyValidatorTest.java:300]: Added tests for empty set and null

File: src/main/java/org/sagebionetworks/bridge/services/StudyService.java (hunk 1 above)
Comment (jnelsonjava) [reply src/main/java/org/sagebionetworks/bridge/services/StudyService.java]: Added null check on scheduleGuid before calling getSchedule.

File: src/main/java/org/sagebionetworks/bridge/services/StudyService.java (hunk 2 above)
Comment (jnelsonjava) [reply src/main/java/org/sagebionetworks/bridge/services/StudyService.java]: Added colon to "custom:"

File: src/main/java/org/sagebionetworks/bridge/validators/StudyValidator.java (hunk 3 above)
Comment (jnelsonjava) [reply src/main/java/org/sagebionetworks/bridge/validators/StudyValidator.java]: Changed to set comparison operations.

File: src/main/java/org/sagebionetworks/bridge/services/StudyService.java [hunk 8]
```diff
@@ -350,4 +355,22 @@ private Study phaseTransition(String appId, String studyId, StudyPhase targetPha
...
         return study;
     }
+    
+    /**
+     * Retrieves the custom event IDs from a schedule. The returned event IDs will 
+     * have the "custom:" prefix removed. If scheduleGuid is null, will return
+     * an empty set.
+     */
+    protected Set<String> getCustomEventIdsFromSchedule(String appId, String scheduleGuid) {
+        Set<String> existingEventIds = new HashSet<>();
```
Comment (alxdarksage) [inline src/main/java/org/sagebionetworks/bridge/services/StudyService.java:365]: very minor: could use ImmutableSet.of() here to avoid creating an object instance

File: src/main/java/org/sagebionetworks/bridge/validators/StudyValidator.java [hunk 9]
```diff
@@ -135,16 +141,21 @@ public void validate(Object object, Errors errors) {
...
             errors.popNestedPath();
         }
         if (uniqueIds.size() > 0 && (uniqueIds.size() != study.getCustomEvents().size())) {
-            errors.rejectValue(CUSTOM_EVENTS_FIELD, "cannot contain duplidate event IDs");
+            errors.rejectValue(CUSTOM_EVENTS_FIELD, "cannot contain duplicate event IDs");
+        }
+        if (protectedCustomEventIds != null && !uniqueIds.containsAll(protectedCustomEventIds)) {
+            protectedCustomEventIds.removeAll(uniqueIds);
+            errors.rejectValue(CUSTOM_EVENTS_FIELD, 
+                    String.format("cannot remove custom events currently used in a schedule: [%s]",
```
Comment (alxdarksage) [inline src/main/java/org/sagebionetworks/bridge/validators/StudyValidator.java]: Minor: you can remove the brackets in the string

Comment (jnelsonjava) [reply src/main/java/org/sagebionetworks/bridge/validators/StudyValidator.java]: Sounds good, done.

File: src/main/java/org/sagebionetworks/bridge/services/StudyService.java (hunk 8 above)
Comment (jnelsonjava) [reply src/main/java/org/sagebionetworks/bridge/services/StudyService.java:365]: I have to leave it as mutable because the validator removes items from the set.
